tests/Unit/Helpers/DataStructures/Tensor/Expressions/EvaluateRank2TestHelpers.hpp

```diff
@@ -0,0 +1,177 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <iterator>
+#include <numeric>
+
+#include "DataStructures/Tensor/Expressions/Evaluate.hpp"
+#include "DataStructures/Tensor/Expressions/TensorExpression.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \ingroup TestingFrameworkGroup
+/// \brief Test that evaluating a right hand side tensor expression containing a
+/// single rank 2 tensor correctly assigns the data to the evaluated left hand
+/// side tensor
+///
+/// \details `TensorIndexA` and `TensorIndexB` can be any type of TensorIndex
+/// and are not necessarily `ti_a_t` and `ti_b_t`. The "A" and "B" suffixes just
+/// denote the ordering of the generic indices of the RHS tensor expression. In
+/// the RHS tensor expression, it means `TensorIndexA` is the first index used
+/// and `TensorIndexB` is the second index used.
+///
+/// If we consider the RHS tensor's generic indices to be (a, b), then this test
+/// checks that the data in the evaluated LHS tensor is correct according to the
+/// index orders of the LHS and RHS. The two possible cases that are checked are
+/// when the LHS tensor is evaluated with index order (a, b) and when it is
+/// evaluated with the index order (b, a).
+///
+/// \tparam Datatype the type of data being stored in the Tensors
+/// \tparam RhsSymmetry the ::Symmetry of the RHS Tensor
+/// \tparam RhsTensorIndexTypeList the RHS Tensor's typelist of
+/// \ref SpacetimeIndex "TensorIndexType"s
+/// \tparam TensorIndexA the type of the first TensorIndex used on the RHS of
+/// the TensorExpression, e.g. `ti_a_t`
+/// \tparam TensorIndexB the type of the second TensorIndex used on the RHS of
+/// the TensorExpression, e.g. `ti_B_t`
+/// \param tensorindex_a the first TensorIndex used on the RHS of the
+/// TensorExpression, e.g. `ti_a`
+/// \param tensorindex_b the second TensorIndex used on the RHS of the
+/// TensorExpression, e.g. `ti_B`
+template <typename Datatype, typename RhsSymmetry,
+          typename RhsTensorIndexTypeList, typename TensorIndexA,
+          typename TensorIndexB>
+void test_evaluate_rank_2_core(
+    const TensorIndexA& tensorindex_a, const TensorIndexB& tensorindex_b) {
+  Tensor<Datatype, RhsSymmetry, RhsTensorIndexTypeList> R_ab{};
+  std::iota(R_ab.begin(), R_ab.end(), 0.0);
```

Comment:
I'm not sure this initialization is correct for `DataType == DataVector`. You need to construct the `DataVector`s with a certain size. We usually do this using
```cpp
auto R_ab = make_with_value<Tensor<Datatype, RhsSymmetry, RhsTensorIndexTypeList>>(0.0, used_for_size);
```
where `used_for_size` is a `DataType` object. Eg a `double` or `DataVector` of the size you want. Nowadays you might be able to just do:
```cpp
auto R_ab = make_with_value<Tensor<Datatype, RhsSymmetry, RhsTensorIndexTypeList>>(0.0, 5_st);
```
and include `Utilities/Literals.hpp`

This applies everywhere you construct tensors by default

Reply:
@nilsdeppe Thanks! I saw some changes with `make_with_value` and `Tensor.hpp` in the last couple weeks, so I took care to rebase before I tried this. I tried rebasing and then replacing this line:

```
Tensor<Datatype, RhsSymmetry, RhsTensorIndexTypeList> R_ab{};
```

with

```
auto R_ab = make_with_value<Tensor<Datatype, RhsSymmetry, RhsTensorIndexTypeList>>(0.0, 5_st);
```

When I do this, [this assert](https://github.com/sxs-collaboration/spectre/blob/49b82942457aba8deb01f65edc2e3667841c4fed/src/Utilities/MakeWithValue.hpp#L31) fails. I then dug around in `Test_Tensor.cpp` for non-default `Tensor` instantiation with `DataVector`s and saw [this](https://github.com/sxs-collaboration/spectre/blob/49b82942457aba8deb01f65edc2e3667841c4fed/tests/Unit/DataStructures/Tensor/Test_Tensor.cpp#L1056) (for example). So, when I instead swapped out the original line for this:

```
Tensor<Datatype, RhsSymmetry, RhsTensorIndexTypeList> R_ab(5_st);
```

it compiled fine and passed the tests. I added a couple new tests (locally) with `DataVector`s and those pass, as well, such as:

```
TestHelpers::TensorExpressions::test_evaluate_rank_2_no_symmetry<
      DataVector, SpacetimeIndex, SpacetimeIndex, UpLo::Lo, UpLo::Up>(
          ti_f, ti_G);
```

Would this change be alright, instead? From my understanding, this would mean that when `Datatype == DataVector` these two lines (the first one changed from the original):

```
Tensor<Datatype, RhsSymmetry, RhsTensorIndexTypeList> R_ab(5_st);
std::iota(R_ab.begin(), R_ab.end(), 0.0);
```

would make the elements of `R_ab` be size 5 `DataVector`s, where the first `DataVector` is `(0, 0, 0, 0, 0)`, the second is `(1, 1, 1, 1, 1)`, the third `(2, 2, 2, 2, 2)`, and so forth. The only comment I have on this is that I don't know how much it matters to have a test where the elements of an individual `DataVector` are not all the same value, as they would be with this suggested change.

Reply:
Yep, sounds good to me! :)